traits/tests/test_ui_notifiers.py

```diff
@@ -61,88 +50,162 @@ def on_foo_change(self, obj, name, old, new):
+        self.addCleanup(trait_notifiers.pop_exception_handler)
+
+    def enterContext(self, cm):
+        # Backport of Python 3.11's TestCase.enterContext method.
+        result = type(cm).__enter__(cm)
+        self.addCleanup(type(cm).__exit__, cm, None, None, None)
+        return result
 
     #### 'TestUINotifiers' protocol ###########################################
 
-    def flush_event_loop(self):
-        """ Post and process the Qt events. """
-        qt4_app.sendPostedEvents()
-        qt4_app.processEvents()
+    def handle_exception(self, object, name, old, new):
+        self.exceptions.append((object, name, old, new))
+
+    def modify_obj(self):
+        trait_notifiers.push_exception_handler(self.handle_exception)
+        try:
+            self.obj.foo = 3
+        finally:
+            trait_notifiers.pop_exception_handler()
 
     def on_foo_notifications(self, obj, name, old, new):
-        thread_id = threading.current_thread().ident
-        event = (thread_id, (obj, name, old, new))
+        event = (threading.current_thread(), (obj, name, old, new))
         self.notifications.append(event)
+        self.done.set()
 
     #### Tests ################################################################
 
-    @unittest.skipIf(
-        not QT_FOUND, "Qt event loop not found, UI dispatch not possible."
-    )
-    def test_notification_from_main_thread(self):
-
-        obj = self.obj_factory()
+    def test_notification_from_main_thread_with_no_ui_handler(self):
+        # Given
+        self.enterContext(clear_ui_handler())
 
-        obj.foo = 3
-        self.flush_event_loop()
+        # When we set obj.foo to 3 on the main thread.
+        self.modify_obj()
 
-        notifications = self.notifications
-        self.assertEqual(len(notifications), 1)
+        # Then the notification is processed synchronously on the main thread.
+        self.assertEqual(
+            self.notifications,
+            [(threading.main_thread(), (self.obj, "foo", 0, 3))],
+        )
 
-        thread_id, event = notifications[0]
-        self.assertEqual(event, (obj, "foo", 0, 3))
+    def test_notification_from_main_thread_with_registered_ui_handler(self):
+        # Given
+        self.enterContext(use_asyncio_ui_handler(asyncio.get_event_loop()))
 
-        ui_thread = trait_notifiers.ui_thread
-        self.assertEqual(thread_id, ui_thread)
+        # When we set obj.foo to 3 on the main thread.
+        self.modify_obj()
 
-    @unittest.skipIf(
-        not QT_FOUND, "Qt event loop not found, UI dispatch not possible."
-    )
-    def test_notification_from_separate_thread(self):
+        # Then the notification is processed synchronously on the main thread.
+        self.assertEqual(
+            self.notifications,
+            [(threading.main_thread(), (self.obj, "foo", 0, 3))],
+        )
 
-        obj = self.obj_factory()
+    def test_notification_from_separate_thread_failure_case(self):
```

Comment:
I think I understand the context for this test case, but just to be clear: do we accept that any background process unknown to the UI _will_ be able to modify objects, but ideally we also want to know about it?

If so, is this a new edge case that we can just test more easily without Qt, or does this happen often?

Reply:
So the key point of the test (and of the `dispatch="ui"` functionality) is that even though the object is modified off the main thread, the trait change _handlers_ are run in the main UI thread.

That said, this pattern isn't ideal in the first place; my preference is to use something like Traits Futures so that we're not modifying the same object from multiple threads in the first place.

Not sure whether this answers your question or not ... ?

Reply:
Ah, I see - so the main result of this test is that we don't see any notifiers and we observe an exception, since the handlers are always on the main UI tread.

That helps understand the context, yes, thanks!

Reply:
Right - the notifiers don't get executed at all in this case.